### PR TITLE
adding keywords and pseudo ops to langspec

### PR DESCRIPTION
**Pull request by barnjamin**

### CODE REVIEW

File: cmd/opdoc/opdoc.go [hunk 1]
```diff
@@ -287,76 +283,146 @@ func typeString(types []logic.StackType) string {
...
-		return fieldsAndTypes(logic.EcdsaCurves)
+		return "ECDSA"
 	default:
-		return nil, ""
+		return ""
+	}
+}
+
+func fieldGroups() []logic.FieldGroup {
+	return []logic.FieldGroup{
```
Comment (barnjamin) [inline cmd/opdoc/opdoc.go:361]: Should we declare this in the `logic` package?

File: cmd/opdoc/opdoc.go [hunk 2]
```diff
@@ -287,76 +283,146 @@ func typeString(types []logic.StackType) string {
...
+		return "json_ref"
 	case "base64_decode":
-		return fieldsAndTypes(logic.Base64Encodings)
+		return "base64"
 	case "vrf_verify":
-		return fieldsAndTypes(logic.VrfStandards)
+		return "vrf_verify"
 	case "ecdsa_pk_recover", "ecdsa_verify", "ecdsa_pk_decompress":
-		return fieldsAndTypes(logic.EcdsaCurves)
+		return "ECDSA"
```
Comment (barnjamin) [inline cmd/opdoc/opdoc.go:354]: Unusual casing on this one

File: data/transactions/logic/assembler.go [hunk 3]
```diff
@@ -1582,12 +1582,12 @@ func getSpec(ops *OpStream, name string, args []string) (OpSpec, string, bool) {
...
-	"int":  {anyImmediates: OpSpec{Name: "int", Proto: proto(":i"), OpDetails: assembler(asmInt)}},
-	"byte": {anyImmediates: OpSpec{Name: "byte", Proto: proto(":b"), OpDetails: assembler(asmByte)}},
+	"int":  {anyImmediates: OpSpec{Name: "int", Proto: proto(":i"), OpDetails: immediates("i").assembler(asmInt)}},
+	"byte": {anyImmediates: OpSpec{Name: "byte", Proto: proto(":b"), OpDetails: immediates("b").assembler(asmByte)}},
 	// parse basics.Address, actually just another []byte constant
-	"addr": {anyImmediates: OpSpec{Name: "addr", Proto: proto(":b"), OpDetails: assembler(asmAddr)}},
+	"addr": {anyImmediates: OpSpec{Name: "addr", Proto: proto(":b"), OpDetails: immediates("a").assembler(asmAddr)}},
 	// take a signature, hash it, and take first 4 bytes, actually just another []byte constant
-	"method":  {anyImmediates: OpSpec{Name: "method", Proto: proto(":b"), OpDetails: assembler(asmMethod)}},
+	"method":  {anyImmediates: OpSpec{Name: "method", Proto: proto(":b"), OpDetails: immediates("m").assembler(asmMethod)}},
```
Comment (barnjamin) [inline data/transactions/logic/assembler.go]: added immediate similar to others i saw but didn't see anything obvious about how to name them?

File: data/transactions/logic/opcodes.go [hunk 4]
```diff
@@ -698,6 +698,8 @@ var opsByOpcode [LogicVersion + 1][256]OpSpec
 // OpsByName map for each version, mapping opcode name to OpSpec
 var OpsByName [LogicVersion + 1]map[string]OpSpec
 
+var PseudoOps []OpSpec
```
Comment (barnjamin) [inline data/transactions/logic/opcodes.go:736]: Wanted something exported from logic package with pseudo ops

File: cmd/opdoc/opdoc.go [hunk 5]
```diff
@@ -236,44 +236,40 @@ func opsToMarkdown(out io.Writer) (err error) {
...
+	DocExtra          string   `json:",omitempty"`
+	ImmediateNote     string   `json:",omitempty"`
+	IntroducedVersion uint64   `json:",omitempty"`
+	Groups            []string `json:",omitempty"`
+}
+
+type Keyword struct {
+	Name      string
+	Type      string
+	ValueEnum []string `json:",omitempty"`
```
Comment (barnjamin) [inline cmd/opdoc/opdoc.go]: Nothing written to use this yet, going back and forth on whether we should embed the TxnType/OnCompletes here or at the same level as other keyword groups.

**Comment by fergalwalsh** [comment]
This looks like a great improvement. 
Maybe `Fields` instead of `Keywords` though?
Including pseudoOps is awesome too.

**Pull request closed without merging**